### PR TITLE
stats: convert Dynamo stats to use StatName interface

### PR DESCRIPTION
**Pull request by jmarantz**
Description: I started off thinking I would try to knock off a bunch of the places where we compute stats by name in one go, but Dynamo deserves its own PR mostly; there's one unrelated file in it. If this goes in first we should remove the whitelist entry in #7573 and if that goes in first then we should merge that in and remove the entry before checking this one in.

This is another step toward resolving #4196 and enabling submission of #4980.

Risk Level: medium -- could introduce lock contention for stats associated with status-codes and table-names. This PR may need more work because of that. Note that #7008 offers a solution to this problem; it's just not clear yet whether it's worth the cost until we see perf traces.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by zuercher** [comment]
/assign @mattklein123 

Feel free to reassign -- just looks like you're the one of the few people who's touched the logic in here.

**Comment by jmarantz** [comment]
OK this is ready for a look -- I removed dynamo_stats_filter from the whitelist established in #7573 .

Feel free to push back on the potential lookup-by-name in the request-path if you think it's an issue, and we can discuss how to resolve.

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    upstream_rq_time_groups_[i] = pool_.add(fmt::format("upstream_rq_time_{}xx", i));
+  }
+  RequestParser::forEachStatString([this](const std::string& str) {
+    // Thread annotation does not realize this function is only called from the
+    // constructor, so we need to lock the mutex. It's easier to just do that
+    // and it's no real penalty.
+    absl::MutexLock lock(&mutex_);
+    builtin_stat_names_[str] = pool_.add(str);
+  });
+  builtin_stat_names_[""] = Stats::StatName();
```
Comment (jmarantz) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: It probably would be useful to remember in the builtins a few common request-code stats here too, like 200, 404, 503 or something, as we do in source/common/http/codes.cc.

However I don't know what the right set of likely status codes is for dynamo DB.

File: source/extensions/filters/http/dynamo/config.cc [hunk 2]
```diff
@@ -14,9 +15,10 @@ namespace Dynamo {
 Http::FilterFactoryCb
 DynamoFilterConfig::createFilter(const std::string& stat_prefix,
                                  Server::Configuration::FactoryContext& context) {
-  return [&context, stat_prefix](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    callbacks.addStreamFilter(Http::StreamFilterSharedPtr{new Dynamo::DynamoFilter(
-        context.runtime(), stat_prefix, context.scope(), context.dispatcher().timeSource())});
+  auto stats = std::make_shared<DynamoStats>(context.scope(), stat_prefix);
+  return [&context, stat_prefix, stats](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    callbacks.addStreamFilter(Http::StreamFilterSharedPtr{
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/config.cc]: nit: convert to make_shared while you are here.

File: source/extensions/filters/http/dynamo/dynamo_filter.h [hunk 3]
```diff
@@ -68,8 +67,7 @@ class DynamoFilter : public Http::StreamFilter {
   void chargeTablePartitionIdStats(const Json::Object& json_body);
 
   Runtime::Loader& runtime_;
-  std::string stat_prefix_;
-  Stats::Scope& scope_;
+  DynamoStatsSharedPtr stats_;
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_filter.h]: nit: const

File: source/extensions/filters/http/dynamo/dynamo_request_parser.h [hunk 4]
```diff
@@ -97,6 +97,17 @@ class RequestParser {
    */
   static std::vector<PartitionDescriptor> parsePartitions(const Json::Object& json_data);
 
+  using StringFn = std::function<void(const std::string)>;
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_request_parser.h]: Pass string by reference or drop the const?

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      capacity_(pool_.add("capacity")), empty_response_body_(pool_.add("empty_response_body")),
+      error_(pool_.add("error")), invalid_req_body_(pool_.add("invalid_req_body")),
+      invalid_resp_body_(pool_.add("invalid_resp_body")),
+      multiple_tables_(pool_.add("multiple_tables")), no_table_(pool_.add("no_table")),
+      operation_missing_(pool_.add("operation_missing")), table_(pool_.add("table")),
+      table_missing_(pool_.add("table_missing")), upstream_rq_time_(pool_.add("upstream_rq_time")),
+      upstream_rq_total_(pool_.add("upstream_rq_total")) {
+  upstream_rq_total_groups_[0] = pool_.add("upstream_rq_total_unknown");
+  upstream_rq_time_groups_[0] = pool_.add("upstream_rq_time_unknown");
+  for (int i = 1; i < 6; ++i) {
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: nit: comment magic 6 or use a named constant?

File: source/extensions/filters/http/dynamo/dynamo_stats.cc (hunk 1 above)
Comment (mattklein123) [reply source/extensions/filters/http/dynamo/dynamo_stats.cc]: Just add a TODO for now? In general can you add TODO/comments about where you think there might be potential perf issues? IMO it's OK in this code to not have it be optimal for now as long as we have comments on the cases in which we might have issues.

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    absl::MutexLock lock(&mutex_);
+    builtin_stat_names_[str] = pool_.add(str);
+  });
+  builtin_stat_names_[""] = Stats::StatName();
+
+  // TODO(jmarantz): should we also learn some 'time' and 'total' stat-names for 200, 404, 503 etc?
+}
+
+Stats::SymbolTable::StoragePtr DynamoStats::addPrefix(const std::vector<Stats::StatName>& names) {
+  std::vector<Stats::StatName> names_with_prefix{prefix_};
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: perf nit: reserve what you need?

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+/**
+ * Creates the partition id stats string.
+ * The stats format is
+ * "<stat_prefix>table.<table_name>.capacity.<operation>.__partition_id=<partition_id>".
+ * Partition ids and dynamodb table names can be long. To satisfy the string length,
+ * we truncate in two ways:
+ * 1. We only take the last 7 characters of the partition id.
+ * 2. If the stats string with <table_name> is longer than the stats MAX_NAME_SIZE, we will
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: Is this comment still valid? I think all the max size stuff is gone? Also please rewrap the comment below.

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+size_t DynamoStats::groupIndex(uint64_t status) {
+  size_t index = status / 100;
+  if (index >= NumGroupEntries) {
+    index = 0; // status-code 600 or higher is unknown.
+  }
+  return index;
+}
+
+Stats::StatName DynamoStats::getStatName(const std::string& str) {
+  // We have been presented with a string to when composing a stat. The Dynamo
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: can't parse. rephrase?

File: source/extensions/filters/http/dynamo/dynamo_stats.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+class DynamoStats {
+public:
+  DynamoStats(Stats::Scope& scope, const std::string& prefix);
+  Stats::SymbolTable& symbolTable() { return scope_.symbolTable(); }
+
+  Stats::Counter& counter(const std::vector<Stats::StatName>& names);
+  Stats::Histogram& histogram(const std::vector<Stats::StatName>& names);
+
+  /**
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.h]: Can we de-dup this comment and either have it here or in the other place with any relevant fixes?

File: source/extensions/filters/http/dynamo/dynamo_stats.cc (hunk 5 above)
Comment (jmarantz) [reply source/extensions/filters/http/dynamo/dynamo_stats.cc]: I made such a constant but didn't use it. using now. good catch.

File: source/extensions/filters/http/dynamo/dynamo_filter.cc [hunk 10]
```diff
@@ -175,39 +175,34 @@ void DynamoFilter::chargeStatsPerEntity(const std::string& entity, const std::st
   std::chrono::milliseconds latency = std::chrono::duration_cast<std::chrono::milliseconds>(
       time_source_.monotonicTime() - start_decode_);
 
-  std::string group_string =
-      Http::CodeUtility::groupStringForResponseCode(static_cast<Http::Code>(status));
+  size_t group_index = DynamoStats::groupIndex(status);
+  Stats::StatNamePool pool(stats_->symbolTable());
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_filter.cc]: Where is this used?

File: source/extensions/filters/http/dynamo/dynamo_stats.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Dynamo {
+
+class DynamoStats {
+public:
+  DynamoStats(Stats::Scope& scope, const std::string& prefix);
+  Stats::SymbolTable& symbolTable() { return scope_.symbolTable(); }
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.h]: Is this used?

**Comment by jmarantz** [comment]
Thanks for the review & catching some unneeded stuff.

**Pull request merged**